quedex_api/user_stream.py

```diff
@@ -0,0 +1,302 @@
+      }
+    """
+    pass
+
+  def on_error(self, error):
+    """
+    Called when an error with market stream occurs (data parsing, signature verification, webosocket
+    error). This means a serious problem, which should be investigated (cf. on_disconnect).
+
+    :type error: subtype of Exception
+    """
+    pass
+
+  def on_disconnect(self, message):
+    """
+    Called when market stream disconnects cleanly (exchange going down for maintenance, network
+    problem, etc.). The client should reconnect in such a case.
+
+    :param message: string message with reason of the disconnect
+    """
+    pass
+
+
+class UserStream(object):
+  def __init__(self, exchange, trader, nonce_group=5):
+    super(UserStream, self).__init__()
+    self.send_message = None
+    self.user_stream_url = exchange.user_stream_url
+
+    self._quedex_key = exchange.public_key
+    self._trader_key = trader.private_key
+    self._account_id = trader.account_id
+
+    self._listeners = []
+    self._nonce_group = nonce_group
+    self._nonce = None
+    self._initialized = False
+
+  def add_listener(self, listener):
+    self._listeners.append(listener)
+
+  def remove_listener(self, listener):
+    self._listeners.remove(listener)
+
+  def place_order(self, place_order_command):
+    """
+    :param place_order_command: a dict of the following format:
+      {
+        "client_order_id": <integer id unique among orders>,
+        "instrument_id": "<string id of the instrument>",
+        "order_type": "limit",
+        "limit_price": "<decimal as string>",
+        "side": "buy"/"sell",
+        "quantity": <integer>,
+      }
+    """
+    if not self._initialized:
+      raise Exception('UserStream not initialized, wait until UserStreamListener.on_read is called')
+    place_order_command['type'] = 'place_order'
+    self._encrypt_send(self._set_nonce_account_id(place_order_command))
+
+  def cancel_order(self, cancel_order_command):
+    """
+    :param cancel_order_command: a dict of the following format:
+      {
+        "client_order_id": <integer id of the order to cancel>,
+      }
+    """
+    if not self._initialized:
+      raise Exception('UserStream not initialized, wait until UserStreamListener.on_read is called')
+    cancel_order_command['type'] = 'cancel_order'
+    self._encrypt_send(self._set_nonce_account_id(cancel_order_command))
+
+  def modify_order(self, modify_order_command):
+    """
+    :param modify_order_command: a dict of the following format:
+      {
+        "client_order_id": <integer id of the order to modify>,
+        "new_limit_price": "<decimal as string>",
+        "new_quantity": <integer>,
+      }
+    """
+    if not self._initialized:
+      raise Exception('UserStream not initialized, wait until UserStreamListener.on_read is called')
+    modify_order_command['type'] = 'modify_order'
+    self._encrypt_send(self._set_nonce_account_id(modify_order_command))
+
+  def batch(self, order_commands):
```

Comment:
I think `batch([commands])` might be redundant as all styles of programming are achievable with `begin_batch`, `send_batch`. Not insisting, though. Good enough for me.